realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -864,21 +867,216 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFind(
+        } else {
+            // handover the result
+            Row row = (*table)[r];
+            std::unique_ptr <SharedGroup::Handover<Row>> handover = SG(
+                    bgSharedGroupPtr)->export_for_handover(row);
+            return reinterpret_cast<jlong>(handover.release());
+        }
+
+    } CATCH_STD()
+    return 0;
+}
+
+
 JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAll(
     JNIEnv* env, jobject, jlong nativeQueryPtr, jlong start, jlong end, jlong limit)
 {
-    Query* pQuery = Q(nativeQueryPtr);
-    Table* pTable = pQuery->get_table().get();
-    if (!QUERY_VALID(env, pQuery) ||
-        !ROW_INDEXES_VALID(env, pTable, start, end, limit))
+    TR_ENTER()
+    Query* query = Q(nativeQueryPtr);
+    Table* table = query->get_table().get();
+    if (!QUERY_VALID(env, query) ||
+        !ROW_INDEXES_VALID(env, table, start, end, limit))
         return -1;
     try {
-        TableView* pResultView = new TableView( pQuery->find_all(S(start), S(end), S(limit)) );
-        return reinterpret_cast<jlong>(pResultView);
+        TableView* tableView = new TableView( query->find_all(S(start), S(end), S(limit)) );
+        return reinterpret_cast<jlong>(tableView);
     } CATCH_STD()
     return -1;
 }
 
+JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAllWithHandover
+  (JNIEnv *env, jobject, jlong bgSharedGroupPtr, jlong replicationPtr, jlong queryPtr, jlong start, jlong end, jlong limit)
+  {
+      TR_ENTER()
+      try {
+          std::unique_ptr<Query> query = getHandoverQuery(bgSharedGroupPtr, replicationPtr, queryPtr);
+          Table* table = query->get_table().get();
+
+          if (!QUERY_VALID(env, query.get()) ||
+              !ROW_INDEXES_VALID(env, table, start, end, limit)) {
+              return 0;
+          }
+
+          // run the query
+          TableView* tableView = new TableView(query->find_all(S(start), S(end), S(limit)));
+
+          // handover the result
+          std::unique_ptr <SharedGroup::Handover<TableView>> handover = SG(
+                  bgSharedGroupPtr)->export_for_handover(*tableView, MutableSourcePayload::Move);
+          return reinterpret_cast<jlong>(handover.release());
+      } CATCH_STD()
+      return 0;
+  }
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAllSortedWithHandover
+  (JNIEnv *env, jobject, jlong bgSharedGroupPtr, jlong replicationPtr, jlong queryPtr, jlong start, jlong end, jlong limit, jlong columnIndex, jboolean ascending)
+  {
+      TR_ENTER()
+      try {
+          std::unique_ptr<Query> query = getHandoverQuery(bgSharedGroupPtr, replicationPtr, queryPtr);
+          Table* table = query->get_table().get();
+
+          if (!QUERY_VALID(env, query.get()) || !ROW_INDEXES_VALID(env, table, start, end, limit)) {
+              return 0;
+          }
+
+          // run the query
+          TableView* tableView = new TableView( query->find_all(S(start), S(end), S(limit)) );
```

Comment:
Release the tableView in the error cases to avoid leaks. Or simply use unique_ptr?